[READY] Allow system boost and system libclang usage

bstaletic opened the pull request:
The changes needed for `--system-libclang` and `--system-boost` to work on Gentoo are:

- Always passing `-lboost_python` even on python3
- Looking in a ridiculous path (`/usr/lib/llvm/${MAJOR_VERSION}/lib{64,32}/`) to find libclang.so

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/837)
<!-- Reviewable:end -->

puremourning commented:
Seems fine. I guess we can handle supporting a _niche_ distro as one of our maintainers uses it :dagger: 

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

One thing i do wonder is if lsb-release is always installed on gentoo? It sometimes comes from some lsb compatibility package or other.

In any case, if it isn't we just don't detect it, and nothing changes so that's fine

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[cpp/CMakeLists.txt, line 163 at r1](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuL4FwDPP1cF_60KkuY:-KuL4FwDPP1cF_60KkuZ:b9huasz) ([raw file](https://github.com/valloric/ycmd/blob/9778bad0b1523a3d888ffb19379be425fe43f43e/cpp/CMakeLists.txt#L163)):*
> ```CMake
> 
> #############################################################################
> # Platform-specific advice goes here. In particular, we have plenty of users
> ```

I don't feel this comment is all that relevant to the code it is attached to.

It made sense where it used to be because that's why we had the validation, and the afforementioned advice. Perhaps it just needs reworking to say that we detect the Linux distribution and do some special stuff.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/837#-:-KuL58QQ107txWMeD1ni:bu31qob)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/837)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
> I guess we can handle supporting a niche distro as one of our maintainers uses it 

I asked on gitter if this PR would be welcome for this exact reason. I won't be against closing this PR if you guys feel Gentoo is too much of a niche distro.

> i do wonder is if lsb-release is always installed on gentoo?

Gentoo doesn't necessarily have `lsb_release` installed. It would be better if I could make cmake parse `/etc/os-release` and have it read just the first line for example. But I have no idea how to parse/source a file with cmake.

For reference, this is what the `/etc/os-release` looks like:

```
NAME=Gentoo
ID=gentoo
PRETTY_NAME="Gentoo/Linux"
ANSI_COLOR="1;32"
HOME_URL="https://www.gentoo.org/"
SUPPORT_URL="https://www.gentoo.org/support/"
BUG_REPORT_URL="https://bugs.gentoo.org/"
```

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[cpp/CMakeLists.txt, line 163 at r1](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuL4FwDPP1cF_60KkuY:-KuL7pqdOyMDb4D2nxP7:b-1vhruj) ([raw file](https://github.com/valloric/ycmd/blob/9778bad0b1523a3d888ffb19379be425fe43f43e/cpp/CMakeLists.txt#L163)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

I don't feel this comment is all that relevant to the code it is attached to.

It made sense where it used to be because that's why we had the validation, and the afforementioned advice. Perhaps it just needs reworking to say that we detect the Linux distribution and do some special stuff.
</blockquote></details>

I'll get to rewriting the comment in a second.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/837#-:-KuL66jWRnvzY-Wso8vB:b-35dkof)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
The codecov is reporting gibberish because of a force push.

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[cpp/CMakeLists.txt, line 163 at r1](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuL4FwDPP1cF_60KkuY:-KuL9hjrFXlUAPepuit_:br4u4bl) ([raw file](https://github.com/valloric/ycmd/blob/9778bad0b1523a3d888ffb19379be425fe43f43e/cpp/CMakeLists.txt#L163)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

I'll get to rewriting the comment in a second.
</blockquote></details>

I just deleted the first line as it seemed the most fitting way to fix the comment. If anyone thinks otherwise, I'd be happy to change it again.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/837#-:-KuL9wBGrOixrGtUcGYk:b-ji40p9)*
<!-- Sent from Reviewable.io -->

puremourning commented:
> I asked on gitter if this PR would be welcome for this exact reason. I won't be against closing this PR if you guys feel Gentoo is too much of a niche distro.

Hehe. I was just joking :) 


And yes, CMake isn't part of codecov, so it's talking nonesense :)

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[cpp/CMakeLists.txt, line 163 at r1](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuL4FwDPP1cF_60KkuY:-KuLIeECjmmkMvWjC90o:b6nts1p) ([raw file](https://github.com/valloric/ycmd/blob/9778bad0b1523a3d888ffb19379be425fe43f43e/cpp/CMakeLists.txt#L163)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

I just deleted the first line as it seemed the most fitting way to fix the comment. If anyone thinks otherwise, I'd be happy to change it again.
</blockquote></details>

Yep, seems fine to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/837#-:-KuLIJl3uD1f6n8-zRay:b-mcjrwf)*
<!-- Sent from Reviewable.io -->

micbou commented:
> Gentoo doesn't necessarily have lsb_release installed. It would be better if I could make cmake parse /etc/os-release and have it read just the first line for example. But I have no idea how to parse/source a file with cmake.

We could do this:
```cmake
if ( EXISTS /etc/os-release )
  file( READ /etc/os-release CONTENTS )
  if ( CONTENTS MATCHES "Gentoo" )
    set( GENTOO_LINUX TRUE )  
  endif()
endif()
```
This can easily be extended to other distributions. For Red Hat and CentOS, we could check if the `/etc/redhat-release` file contains `Red Hat` (respectively `CentOS`).

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cpp/CMakeLists.txt, line 164 at r2](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuLQf49H3CcYunyI7tV:-KuLQf4ARAjK6fZ_grrH:b-xpif2u) ([raw file](https://github.com/valloric/ycmd/blob/a0d170431fa3066bf9d989a6f69fcf137b6da9d8/cpp/CMakeLists.txt#L164)):*
> ```CMake
> #############################################################################
> 
> # We use lsb_release on linux systems to detect the distro/verseion.
> ```

Typo in `version`.

---

*[cpp/ycm/CMakeLists.txt, line 285 at r2](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuLPytomUj_h-1b_ROW:-KuLPytomUj_h-1b_ROX:b-yw48y6) ([raw file](https://github.com/valloric/ycmd/blob/a0d170431fa3066bf9d989a6f69fcf137b6da9d8/cpp/ycm/CMakeLists.txt#L285)):*
> ```CMake
>     file ( GLOB FREEBSD_LLVM_PATHS "/usr/local/llvm*/lib")
>     # On Gentoo, llvm installs into /usr/lib/llvm/x/
>     file ( GLOB GENTOO_LLVM_PATHS "/usr/lib/llvm/*/lib*" )
> ```

You said on Gitter that LLVM is placed in the `/usr/lib64` folder. Should we check for both folders (`/usr/lib` and `/usr/lib64`)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/837#-:-KuLO4n9kdwqya5jsbWJ:b-ryjk0m)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Thanks, @micbou. I switched to `os-release`parsing instead of `lsb_release`.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[cpp/CMakeLists.txt, line 163 at r1](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuL4FwDPP1cF_60KkuY:-KuLU_TA7r8c_VQGKSon:bcukx72) ([raw file](https://github.com/valloric/ycmd/blob/9778bad0b1523a3d888ffb19379be425fe43f43e/cpp/CMakeLists.txt#L163)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Yep, seems fine to me.
</blockquote></details>

I reverted the whole block to the way it were, because I went for `/etc/os-release` parsing to make sure we detec Gentoo.

---

*[cpp/CMakeLists.txt, line 164 at r2](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuLQf49H3CcYunyI7tV:-KuLV9VnwVjEkc6XBUpY:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/a0d170431fa3066bf9d989a6f69fcf137b6da9d8/cpp/CMakeLists.txt#L164)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Typo in `version`.
</blockquote></details>

Done.

---

*[cpp/ycm/CMakeLists.txt, line 285 at r2](https://reviewable.io:443/reviews/valloric/ycmd/837#-KuLPytomUj_h-1b_ROW:-KuLSizlS8XA1uQJBkej:b65vqs5) ([raw file](https://github.com/valloric/ycmd/blob/a0d170431fa3066bf9d989a6f69fcf137b6da9d8/cpp/ycm/CMakeLists.txt#L285)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

You said on Gitter that LLVM is placed in the `/usr/lib64` folder. Should we check for both folders (`/usr/lib` and `/usr/lib64`)?
</blockquote></details>

`/usr/lib` is a symlink to `/usr/lib64` on my machine, but I can't guarantee that's thte case for every Gentoo.
Especially since I know there are people there with static `/dev` and even persistent `/run`.
Checking both seems like a good idea.

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/837#-:-KuLSGVXfUfqFK6MJ39u:bb8ejre)*
<!-- Sent from Reviewable.io -->

micbou commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

@zzbot r+

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/837#-:-KuM6BjBoMSaENHc2iw2:b-a2cmdb)*
<!-- Sent from Reviewable.io -->
